### PR TITLE
Mailbox 278 Make sure indexing index at least email body when attachment conversion fails

### PR DESCRIPTION
**Pull request by Laura-Royet**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 1]
```diff
@@ -2465,4 +2465,173 @@ private boolean messageInMailboxHasHeaders(String mailboxId, String messageId, I
...
+        String messageCreationId = "creationId";
+        String fromAddress = username;
+        String outboxId = getOutboxId(accessToken);
+        String requestBody = "[" +
+                "  [" +
+                "    \"setMessages\","+
+                "    {" +
+                "      \"create\": { \"" + messageCreationId  + "\" : {" +
+                "        \"from\": { \"name\": \"Me\", \"email\": \"" + fromAddress + "\"}," +
+                "        \"to\": [{ \"name\": \"BOB\", \"email\": \"" + fromAddress + "\"}]," +
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: can you replace BOB by Me as it this the same address?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 2]
```diff
@@ -2465,4 +2465,173 @@ private boolean messageInMailboxHasHeaders(String mailboxId, String messageId, I
...
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 3]
```diff
@@ -2465,4 +2465,173 @@ private boolean messageInMailboxHasHeaders(String mailboxId, String messageId, I
...
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: too

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 4]
```diff
@@ -2465,4 +2465,173 @@ private boolean messageInMailboxHasHeaders(String mailboxId, String messageId, I
...
+        String presumedMessageId = "username@domain.tld|INBOX|1";
+        
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body("[[\"getMessageList\", {\"filter\":{\"body\": \"Test\"}}, \"#0\"]]")
+            .when()
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .log().all()//ifValidationFails()
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: please keep the ifValidationFails()

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java [hunk 5]
```diff
@@ -103,10 +103,14 @@ public void add(MailboxSession session, Mailbox mailbox, MailboxMessage message)
         try {
             indexer.indexMessage(indexIdFor(mailbox, message.getUid()), messageToElasticSearchJson.convertToJson(message, ImmutableList.of(session.getUser())));
         } catch (Exception e) {
-            LOGGER.error("Error when indexing message " + message.getUid(), e);
+            try {
+                indexer.indexMessage(indexIdFor(mailbox, message.getUid()), messageToElasticSearchJson.convertToJsonWithoutAttachment(message, ImmutableList.of(session.getUser())));
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java]: Maybe we can have a info log here ?

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 6]
```diff
@@ -89,9 +90,47 @@ public void addShouldIndex() throws Exception {
         Mailbox mailbox = control.createMock(Mailbox.class);
         MessageUid messageId = MessageUid.of(1);
         TestId mailboxId = TestId.of(12);
-        expect(mailbox.getMailboxId()).andReturn(mailboxId);
+        expect(mailbox.getMailboxId())
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: Time to switch this to Mockito we use everywhere else ?

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java [hunk 7]
```diff
@@ -103,10 +103,14 @@ public void add(MailboxSession session, Mailbox mailbox, MailboxMessage message)
...
+            } catch (JsonProcessingException e1) {
+                LOGGER.error("Error when basic indexing of message" + message.getUid(), e1);
```
Comment (Laura-Royet) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java]: Done

**Comment by aduprat** [comment]
I think the catch is at the wrong place, it should be in TikaTextExtractor.

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 8]
```diff
@@ -89,9 +90,47 @@ public void addShouldIndex() throws Exception {
...
+            .andReturn(mailboxId);
         MailboxMessage message = mockedMessage(messageId);
         
+        expect(messageToElasticSearchJson.convertToJson(anyObject(MailboxMessage.class), anyObject()))
```
Comment (Laura-Royet) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: done

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 9]
```diff
@@ -89,9 +90,47 @@ public void addShouldIndex() throws Exception {
...
+        Mailbox mailbox = control.createMock(Mailbox.class);
+        MessageUid messageId = MessageUid.of(1);
+        TestId mailboxId = TestId.of(12);
+        expect(mailbox.getMailboxId())
+            .andReturn(mailboxId)
+            .times(2);
+        MailboxMessage message = mockedMessage(messageId);
+        
+        JsonProcessingException expectedJsonProcessingException = control.createMock(JsonProcessingException.class);
+        expect(messageToElasticSearchJson.convertToJson(eq(message), anyObject()))
```
Comment (Laura-Royet) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: ok

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 10]
```diff
@@ -112,17 +153,25 @@ public void addShouldNotPropagateExceptionWhenExceptionOccurs() throws Exception
...
+        expect(mailbox.getMailboxId())
+            .andReturn(mailboxId)
+            .times(2);
         MailboxMessage message = mockedMessage(messageId);
-        expect(mailbox.getMailboxId()).andReturn(mailboxId);
         
-        expect(indexer.indexMessage(eq(mailboxId.serialize() + ":" + messageId), anyString()))
-            .andThrow(new ElasticsearchException(""));
+        JsonProcessingException expectedJsonProcessingException = control.createMock(JsonProcessingException.class);
+        expect(messageToElasticSearchJson.convertToJson(eq(message), anyObject()))
```
Comment (Laura-Royet) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 1 above)
Comment (Laura-Royet) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: ok :)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 2 above)
Comment (Laura-Royet) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 3 above)
Comment (Laura-Royet) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 4 above)
Comment (Laura-Royet) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: ok

**Comment by Laura-Royet** [comment]
test this please

**Comment by Laura-Royet** [comment]
About chibenwa  following comment "You should allow partial attachments success : If attachment convertion A fails, but B success, we WANT TO index B..."
Response: this PR is about ES indexing failure. When it occurs the only indexed part of the message is its email body as asked in this issue. 
What was your idea chibenwa?

**Comment by chibenwa** [comment]
No it's a tika issue.

Please put catch in Tika.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIndexer.java [hunk 11]
```diff
@@ -53,6 +55,29 @@ public String getId() {
         public String getUpdatedDocumentPart() {
             return updatedDocumentPart;
         }
+
+        @Override
+        public final boolean equals(Object o) {
```
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIndexer.java]: Optimisation suggestion (but I think this is small), start with:
`if (this == o) {
   return true; 
 }`

Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIndexer.java]: It's a known one but the added complexity is not worth it (at least it's the implicit team choice we've made).

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 12]
```diff
@@ -18,315 +18,312 @@
...
-            .andThrow(expectedJsonProcessingException)
-            .once();
-        expect(messageToElasticSearchJson.convertToJsonWithoutAttachment(anyObject(MailboxMessage.class), anyObject()))
-            .andReturn("json content")
-            .once();
-
-        IndexResponse expectedIndexResponse = control.createMock(IndexResponse.class);
-        expect(indexer.indexMessage(eq(mailboxId.serialize() + ":" + messageId), anyString()))
-            .andReturn(expectedIndexResponse);
+        String expectedJsonContent = "json content";
```
Comment (rouazana) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: Should be a constant

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 13]
```diff
@@ -18,315 +18,312 @@
...
+                .thenReturn(mailboxId);
+            
+            when(messageToElasticSearchJson.getUpdatedJsonMessagePart(any(Flags.class), any(Long.class)))
+                .thenReturn("json updated content");
+            
+            //When
+            testee.update(session, mailbox, Lists.newArrayList(updatedFlags));
+            
+            //Then
+            ImmutableList<UpdatedRepresentation> expectedUpdatedRepresentations = ImmutableList.of(new UpdatedRepresentation(mailboxId.serialize() + ":" + messageId.asLong(), "json updated content"));
```
Comment (rouazana) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: you can replace the content of UpdatedRepresentation by its actual value

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+            Mailbox mailbox = mock(Mailbox.class);
+            Flags flags = new Flags();
+            MessageUid messageId = MessageUid.of(1);
+            UpdatedFlags updatedFlags = new UpdatedFlags(messageId, MODSEQ, flags, flags);
+            TestId mailboxId = TestId.of(12);
+    
+            when(mailbox.getMailboxId())
+                .thenReturn(mailboxId);
+
+            ImmutableList<UpdatedRepresentation> expectedUpdatedRepresentations = ImmutableList.of(new UpdatedRepresentation(mailboxId.serialize() + ":" + messageId.asLong(), "json updated content"));
```
Comment (rouazana) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: Still this one to fix

Comment (Laura-Royet) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: Thanks, done.

**Comment by Laura-Royet** [comment]
test this please

**Comment by Laura-Royet** [comment]
These new fixup was made to solve the 2 integration tests that failed.

File: mailbox/store/src/test/resources/eml/emailWithNonIndexableAttachment.eml [hunk 15]
```diff
@@ -0,0 +1,1655 @@
...
+Subject: test
+Message-ID: 1480502737913
+Date: Wed, 30 Nov 2016 10:45:38 +0000
+
+---=Part.0.16b49c355c1c1b54.158b4d6fd1b.bd3741f90275bb0a=-
+Content-Type: text/html; name="=?US-ASCII?Q?test.html?="
+Content-Disposition: attachment
+Content-Transfer-Encoding: base64
+
+PCFET0NUWVBFIGh0bWw+CjxodG1sIGNsYXNzPSJtb3ppbGxhIiBsYW5nPSJlbiI+PGhlYWQ+Cjxt
```
Comment (chibenwa) [inline mailbox/store/src/test/resources/eml/emailWithNonIndexableAttachment.eml]: Could we try to reduce this hudge diff on this file ? 1000 line + seems a bit to much....

File: mailbox/store/src/test/resources/eml/emailWithNonIndexableAttachment.eml [hunk 16]
```diff
@@ -0,0 +1,1655 @@
...
+ZXRhIGh0dHAtZXF1aXY9ImNvbnRlbnQtdHlwZSIgY29udGVudD0idGV4dC9odG1sOyBjaGFyc2V0
```
Comment (chibenwa) [inline mailbox/store/src/test/resources/eml/emailWithNonIndexableAttachment.eml]: Could we check that we support encoding base64 for indexing ? (that might be the underlying issue, no ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/attachment/nonIndexableAttachment.html [hunk 17]
```diff
@@ -0,0 +1,827 @@
...
...-edit-operation" data-operation="(i)"><img src="test_files/information.gif" alt="information"></a></li><li><a original-title="(/)" href="#" class="wiki-edit-operation" data-operation="(/)"><img src="test_files/check.gif" alt="check"></a></li><li><a original-title="(x)" href="#" class="wiki-edit-operation" data-operation="(x)"><img src="test_files/error.gif" alt="error"></a></li><li><a original-title="(!)" href="#" class="wiki-edit-operation" data-operation="(!)"><img src="test_files/warning.gif" alt="warning"></a></li><li><a original-title="(+)" href="#" class="wiki-edit-operation" data-operation="(+)"><img src="test_files/add.gif" alt="add"></a></li><li><a original-title="(-)" href="#" class="wiki-edit-operation" data-operation="(-)"><img src="test_files/forbidden.gif" alt="forbidden"></a></li><li><a original-title="(?)" href="#" class="wiki-edit-operation" data-operation="(?)"><img src="test_files/help_16.gif" alt="help"></a></li><li><a original-title="(on)" href="#" class="wiki-edit-operation" data-operation="(on)"><img src="test_files/lightbulb_on.gif" alt="lightbulb on"></a></li><li><a original-title="(off)" href="#" class="wiki-edit-operation" data-operation="(off)"><img src=
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/attachment/nonIndexableAttachment.html]: The last line here is not collored by github. Maybe the non indexable attachment is in here....

**Comment by rouazana** [comment]
see #596

**Pull request closed without merging**